CRUD Openstack

dLobatog opened the pull request:
Basically these are some changes to make vms_controller for compute_resources work properly with Openstack so we can suspend/pause/resume an instance (not possible as of now).  Also changed some naming to make it look like Openstack rather than some generic stuff. 

For instance: status of a vm now is Active/Suspended, etc.. instead of On and Off which is a convention that does not make much sense for Openstack.

New: I added support for opening new VNC consoles on Openstack VMs.

dLobatog committed:
Added ability to pause/suspend Openstack instances. VM status is more informative

dLobatog committed:
Proper naming of Openstack VM instances

dLobatog committed:
Script to set up Foreman first install automatically

dLobatog committed:
Added ability to pause/suspend Openstack instances. Loads tenants properly. VM status is more informative

dLobatog committed:
Proper naming of Openstack VM instances

dLobatog committed:
Openstack console support for virtual machines (VNC)

ohadlevy commented on app/controllers/compute_resources_vms_controller.rb:
```diff
@@ -30,17 +30,45 @@ def show
   end
 
   def power
+    (power_openstack and return) if @vm.class == Fog::Compute::OpenStack::Server
```
imho this logic needs to move to lib/fog_extensions/openstack/server.rb

ohadlevy commented on app/models/compute_resource.rb:
```diff
@@ -152,7 +152,12 @@ def as_json(options={})
 
   def console uuid = nil
-    raise "#{provider} console is not supported at this time"
+    raise "#{provider} console is not supported at this time" unless provider == 'Openstack'
```
the whole point of compute resources and the fact that they have subclasses (e.g. ec2, openstack etc) is that you can override the methods :) so you dont need to hack hacks like this.

simply deifne a new console method in the lib/foreman/model/openstack.rb file

dLobatog replied:
That's a good point, sorry. Will update the PR in a matter of minutes with this.

dLobatog committed:
Use Openstack fog extension to extend power/suspend classes

dLobatog committed:
Image form shows a dropdown with images for Openstack rather than a blank text box

dLobatog committed:
Console is overriden in lib/foreman instead of hacked around in compute_resource_vms

dLobatog committed:
Added support to create new Openstack VMs

ohadlevy commented on app/controllers/compute_resources_vms_controller.rb:
```diff
@@ -51,10 +64,20 @@ def destroy
-  rescue => e
-    process_error :redirect => compute_resource_vm_path(@compute_resource, @vm.identity), :error_msg => "Failed to set console: #{e}", :object => @vm
+    if @compute_resource.type == "Foreman::Model::Openstack"
+      openstack_vnc_console
```
since you override the console method, I am not sure if this logic is required?

dLobatog replied:
fixed, I just redirect to the right VNC url unless we get not a string and whatever the console object is supposed to be. 
I'm not sure I get how template hosts/console works, what should the Console object look like? If it's a Excon response like it is when it comes from OpenStack through Fog, I don't think it has any of the parameters used in hosts/console

ohadlevy commented on app/helpers/images_helper.rb:
```diff
@@ -1,2 +1,8 @@
 module ImagesHelper
+  def openstack_list_of_images
+    hash = {}
+    ComputeResource.first.vms.first.connection.list_images.body['images'].each { |image| hash[image['id']] = image['name']}   
```
I think this better move into the openstack compute resource itself, surely CompureResource.first wont always be openstack :)

something like

def avail_images
   client.list_images.body.....
end

probably better to extend fog to provide back images instead of going to fog requests directly.

dLobatog replied:
Okay, I do this in lib/fog_extensions/openstack then.

ohadlevy commented on app/views/compute_resources_vms/index/_openstack.html.erb:
```diff
@@ -13,10 +13,12 @@
       <td><%= vm.flavor_with_object %></td>
       <td><%= vm.tenant %></td>
-      <td <%= vm_power_class(vm.ready?)%>> <%= vm_state(!vm.ready?) %> </td>
+      <td <%= vm_power_class(vm.ready?)%>> <%= vm.state.capitalize %> </td>
```
i need to double check that this works across all compute resources

ohadlevy commented on app/views/images/_form.html.erb:
```diff
@@ -6,7 +6,11 @@
   <%= select_f f, :architecture_id, Architecture.all, :id, :to_label %>
   <%= text_f f, :username, :value => @image.username || "root", :help_inline => "The user that is used to ssh into the instance, normally ec2-user, ubuntu, root etc" %>
+  <% if @compute_resource.type == 'Foreman::Model::Openstack' %>
+  <%= select_f f, :uuid, openstack_list_of_images, :first, :last %>
```
sounds to me this is not open stack specific, so we could probably extend the parent class with all of these attributes.

ohadlevy commented on lib/fog_extensions/openstack/server.rb:
```diff
@@ -8,7 +8,13 @@ def tenant
 
-  #    alias_method_chain :flavor, :object
+      def start
+        connection.resume_server(id)
```
so you can only resume a runner instance? how about an instance that was never powered on?

dLobatog replied:
AFAIK you always power on an instance because when you create it, it powers on itself automatically after build. I don't think this is feasible not even from OpenStack nova (just because it's never needed)

ohadlevy commented:
ping?

dLobatog commented:
Working on this atm.

dLobatog committed:
Fixed compute resources vms merge

dLobatog committed:
Merge branch 'develop' of git://github.com/theforeman/foreman into openstack

dLobatog committed:
Removed specific call to Openstack console and instead I do this on the Openstack compute resource itself

Fix to print OpenStack security groups

Logic to list the images is in fog_extensions/openstack

Logic to list the images is in fog_extensions/openstack

Logic to list the images is in fog_extensions/openstack

Added a link to create a new VM. Only tested for Openstack and it works :)

dLobatog committed:
Removed specific call to Openstack console and instead I do this on the Openstack compute resource itself

Fix to print OpenStack security groups

Logic to list the images is in fog_extensions/openstack

Logic to list the images is in fog_extensions/openstack

Logic to list the images is in fog_extensions/openstack

Added a link to create a new VM. Only tested for Openstack and it works :)

Fixed reference to openstack_list_images

dLobatog commented:
Updated the pull request.

dLobatog committed:
Merge branch 'develop' of git://github.com/theforeman/foreman into openstack

dLobatog committed:
Merge branch 'openstack' of github.com:eLobato/foreman into openstack

ohadlevy commented:
most of the issues here have been merged - please open a new PR if required - thanks!
